cpp/src/arrow/compute/kernels/aggregate_basic_internal.h

```diff
@@ -186,11 +186,8 @@ struct MeanImpl : public SumImpl<ArrowType, SimdLevel> {
       // Round the decimal result based on the remainder
       remainder.Abs();
       if (remainder * 2 >= this->count) {
-        if (this->sum >= 0) {
-          quotient += 1;
-        } else {
-          quotient -= 1;
-        }
+        auto flag = this->sum >= 0;
+        quotient = quotient + (1 & flag) - (1 & (!flag));
```

Comment:
My gut feeling is compiler should be able to do necessary optimization of this code pattern. It may be better to keep current more readable code.
Do you have benchmark result? Or can you write a short test code and compare the machine code on godblot?

Reply:
Looks the original naive c code generates better machine instructions. https://godbolt.org/z/KWrTrWKac

Please note the Arm "add shift" opcode has 2 cycles latency, so the new code is still worse than original one though both contains three machine instructions.

Conditional instructions like `cmov`, `csinc` can easily eliminate this trivial branch.

Reply:
I'm not surprised, this kind of micro-optimization where you try to second-guess actual CPU instructions and their performance using C code can often be an anti-pattern.